File: ios/PouchProv/Sources/PouchProvCore/Messages.swift
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Provisioning message codec (cddl/prov.cddl, protocol version 1).
///
/// Every message is a CBOR array `[op, ...]`. This mirrors the Kotlin
/// `Messages.kt` and the Python `codec.py`, and is pinned to the device's zcbor
/// codec via the shared golden vectors in `tests/vectors/`.
public enum Messages {
    public static let protoVersion = 1

    // Reserved provisioning paths (pouch downlink/uplink entry paths).
    public static let pathVer = ".prov/ver"
    public static let pathAuth = ".prov/auth"
    public static let pathConfig = ".prov/config"
    public static let pathScan = ".prov/scan"
    public static let pathCred = ".prov/cred"
    public static let pathCtrl = ".prov/ctrl"

    // MARK: - .prov/ver

    public static func encodeVerReq() -> Data {
        Cbor.encode(.array([.int(0)]))
    }

    public static func decodeVerRsp(_ data: Data) throws -> VersionInfo {
        let msg = try decodeResponse(data, expectOp: 0, context: "ver")
        guard msg.count > 2, case .map = msg[2],
              let proto = msg[2]["proto"]?.intValue,
              let rawCaps = msg[2]["caps"]?.arrayValue,
              let blockSize = msg[2]["blk"]?.intValue,
              let lib = msg[2]["lib"]?.textValue
        else { throw DecodeError("ver: bad info map") }
        let caps = try rawCaps.map { cap -> String in
            guard let text = cap.textValue else { throw DecodeError("ver: bad cap") }
            return text
        }
        return VersionInfo(
            proto: proto,
            caps: caps,
            blockSize: blockSize,
            lib: lib,
            popRequired: msg[2]["pop"]?.boolValue ?? false
        )
    }

    // MARK: - .prov/auth

    public static func encodeAuthChallenge(cliNonce: Data) -> Data {
        precondition(cliNonce.count == 16, "cli-nonce must be 16 bytes")
        return Cbor.encode(.array([.int(0), .bytes(cliNonce)]))
    }

    /// Returns (devNonce, devProof).
    public static func decodeAuthChallengeRsp(_ data: Data) throws -> (devNonce: Data, devProof: Data) {
        let msg = try decodeResponse(data, expectOp: 0, context: "auth")
        guard msg.count > 3,
              let devNonce = msg[2].bytesValue,
              let devProof = msg[3].bytesValue
        else { throw DecodeError("auth: bad challenge response") }
        guard devNonce.count == 16, devProof.count == 32 else {
            throw DecodeError("auth: bad nonce/proof size")
        }
        return (devNonce, devProof)
    }

    public static func encodeAuthProof(cliProof: Data) -> Data {
        precondition(cliProof.count == 32, "cli-proof must be 32 bytes")
        return Cbor.encode(.array([.int(1), .bytes(cliProof)]))
    }

    public static func decodeAuthProofRsp(_ data: Data) throws {
        _ = try decodeResponse(data, expectOp: 1, context: "auth")
    }

    // MARK: - .prov/config

    public static func encodeConfigGetStatus() -> Data {
        Cbor.encode(.array([.int(0)]))
    }

    public static func encodeConfigSet(
        ssid: Data,
        password: Data? = nil,
        bssid: Data? = nil,
        channel: Int? = nil
    ) -> Data {
        var cfg: [CborPair] = [CborPair(.text("ssid"), .bytes(ssid))]
        if let password { cfg.append(CborPair(.text("pass"), .bytes(password))) }
        if let bssid { cfg.append(CborPair(.text("bssid"), .bytes(bssid))) }
        if let channel { cfg.append(CborPair(.text("ch"), .int(Int64(channel)))) }
        return Cbor.encode(.array([.int(1), .map(cfg)]))
    }

    public static func encodeConfigApply() -> Data {
        Cbor.encode(.array([.int(2)]))
    }

    public static func decodeConfigStatusRsp(_ data: Data) throws -> WifiStatus {
        let msg = try decodeResponse(data, expectOp: 0, context: "config")
        guard msg.count > 2, let stateCode = msg[2].intValue else {
            throw DecodeError("config: bad sta-state")
        }
        let state = try StaState.from(stateCode)
        var failReason: FailReason?
        var ip4: Data?
        var ssid: Data?
        var rssi: Int?
        if msg.count > 3 {
            switch msg[3] {
            case .int(let reason):
                failReason = try FailReason.from(Int(reason))
            case .map:
                guard let detailIp4 = msg[3]["ip4"]?.bytesValue,
                      let detailSsid = msg[3]["ssid"]?.bytesValue
                else { throw DecodeError("config: bad status detail") }
                ip4 = detailIp4
                ssid = detailSsid
                rssi = msg[3]["rssi"]?.intValue
            default:
                throw DecodeError("config: bad status detail")
            }
        }
        return WifiStatus(state: state, failReason: failReason, ip4: ip4, ssid: ssid, rssi: rssi)
    }

    public static func decodeConfigSetRsp(_ data: Data) throws {
        _ = try decodeResponse(data, expectOp: 1, context: "config")
    }

    public static func decodeConfigApplyRsp(_ data: Data) throws {
        _ = try decodeResponse(data, expectOp: 2, context: "config")
    }

    // MARK: - .prov/scan

    public static func encodeScanStart(passive: Bool? = nil, periodMs: Int? = nil) -> Data {
        var params: [CborPair] = []
        if let passive { params.append(CborPair(.text("passive"), .bool(passive))) }
        if let periodMs { params.append(CborPair(.text("period-ms"), .int(Int64(periodMs)))) }
        return Cbor.encode(.array([.int(0), .map(params)]))
    }

    public static func encodeScanGetStatus() -> Data {
        Cbor.encode(.array([.int(1)]))
    }

    public static func encodeScanGetResults(start: Int, count: Int) -> Data {
        Cbor.encode(.array([.int(2), .int(Int64(start)), .int(Int64(count))]))
    }

    public static func decodeScanStartRsp(_ data: Data) throws {
        _ = try decodeResponse(data, expectOp: 0, context: "scan")
    }

    /// Returns (finished, total).
    public static func decodeScanStatusRsp(_ data: Data) throws -> (finished: Bool, total: Int) {
        let msg = try decodeResponse(data, expectOp: 1, context: "scan")
        guard msg.count > 3,
              let finished = msg[2].boolValue,
              let total = msg[3].intValue
        else { throw DecodeError("scan: bad status response") }
        return (finished, total)
    }

    public static func decodeScanResultsRsp(_ data: Data) throws -> [ScanEntry] {
        let msg = try decodeResponse(data, expectOp: 2, context: "scan")
        guard msg.count > 2, let list = msg[2].arrayValue else {
            throw DecodeError("scan: bad results response")
        }
        return try list.map { entry -> ScanEntry in
            guard case .map = entry,
                  let ssid = entry["ssid"]?.bytesValue,
                  let bssid = entry["bssid"]?.bytesValue,
                  let channel = entry["ch"]?.intValue,
                  let rssi = entry["rssi"]?.intValue,
                  let auth = entry["auth"]?.intValue
            else { throw DecodeError("scan: bad result entry") }
            return ScanEntry(ssid: ssid, bssid: bssid, channel: channel, rssi: rssi, auth: auth)
        }
    }

    // MARK: - .prov/cred

    public static func encodeCredWrite(kind: CredKind, offset: Int, total: Int, data: Data) -> Data {
        let map: [CborPair] = [
            CborPair(.text("kind"), .int(Int64(kind.rawValue))),
            CborPair(.text("off"), .int(Int64(offset))),
            CborPair(.text("total"), .int(Int64(total))),
            CborPair(.text("data"), .bytes(data)),
        ]
        return Cbor.encode(.array([.int(0), .map(map)]))
    }

    public static func encodeCredFinalize() -> Data {
        Cbor.encode(.array([.int(1)]))
    }

    public static func encodeCredGetStatus() -> Data {
        Cbor.encode(.array([.int(2)]))
    }

    /// Returns total bytes received for the written kind.
    public static func decodeCredWriteRsp(_ data: Data) throws -> Int {
        let msg = try decodeResponse(data, expectOp: 0, context: "cred")
        guard msg.count > 2, let received = msg[2].intValue else {
            throw DecodeError("cred: bad write response")
        }
        return received
    }

    public static func decodeCredFinalizeRsp(_ data: Data) throws {
        _ = try decodeResponse(data, expectOp: 1, context: "cred")
    }

    public static func decodeCredStatusRsp(_ data: Data) throws -> [CredKind: Int] {
        let msg = try decodeResponse(data, expectOp: 2, context: "cred")
        guard msg.count > 2, let pairs = msg[2].mapValue else {
            throw DecodeError("cred: bad status response")
        }
        var status: [CredKind: Int] = [:]
        for pair in pairs {
            guard let key = pair.key.textValue, let code = Int(key), let len = pair.value.intValue else {
                throw DecodeError("cred: bad status entry")
            }
            status[try CredKind.from(code)] = len
        }
        return status
    }

    // MARK: - .prov/ctrl

    public static func encodeCtrl(_ op: CtrlOp) -> Data {
        Cbor.encode(.array([.int(Int64(op.rawValue))]))
    }

    public static func decodeCtrlRsp(_ data: Data, op: CtrlOp) throws {
        _ = try decodeResponse(data, expectOp: op.rawValue, context: "ctrl")
    }

    // MARK: - shared

    private static func decodeResponse(_ data: Data, expectOp: Int, context: String) throws -> [CborValue] {
        let obj: CborValue
        do {
            obj = try Cbor.decode(data)
        } catch {
            throw DecodeError("\(context): bad CBOR: \(error)")
        }
        guard let msg = obj.arrayValue, msg.count >= 2 else {
            throw DecodeError("\(context): not a response array")
        }
        guard let op = msg[0].intValue, op == expectOp else {
            throw DecodeError("\(context): op \(msg[0]), expected \(expectOp)")
        }
        guard let statusCode = msg[1].intValue else {
            throw DecodeError("\(context): bad status")
        }
        let status = try Status.from(statusCode)
        guard status == .ok else { throw ProvError(status: status, context: context) }
        return msg
    }
}

// MARK: - enums

public enum Status: Int, Equatable, Sendable {
    case ok = 0
    case invalidProto = 1
    case invalidArgument = 2
    case internalError = 3
    case unauthorized = 4
    case invalidState = 5
    case busy = 6

    static func from(_ code: Int) throws -> Status {
        guard let status = Status(rawValue: code) else { throw DecodeError("unknown status \(code)") }
        return status
    }
}

public enum StaState: Int, Equatable, Sendable {
    case connected = 0
    case connecting = 1
    case disconnected = 2
    case failed = 3

    static func from(_ code: Int) throws -> StaState {
        guard let state = StaState(rawValue: code) else { throw DecodeError("unknown sta-state \(code)") }
        return state
    }
}

public enum FailReason: Int, Equatable, Sendable {
    case authError = 0
    case networkNotFound = 1

    static func from(_ code: Int) throws -> FailReason {
        guard let reason = FailReason(rawValue: code) else { throw DecodeError("unknown fail-reason \(code)") }
        return reason
    }
}

public enum CredKind: Int, Equatable, Sendable {
    case deviceCert = 0
    case privateKey = 1
    case caCert = 2

    static func from(_ code: Int) throws -> CredKind {
        guard let kind = CredKind(rawValue: code) else { throw DecodeError("unknown cred kind \(code)") }
        return kind
    }
}

public enum CtrlOp: Int, Equatable, Sendable {
    case reset = 0
    case reprovision = 1
    case end = 2
}

// MARK: - value types

public struct VersionInfo: Equatable, Sendable {
    public let proto: Int
    public let caps: [String]
    public let blockSize: Int
    public let lib: String
    public let popRequired: Bool

    public init(proto: Int, caps: [String], blockSize: Int, lib: String, popRequired: Bool) {
        self.proto = proto
        self.caps = caps
        self.blockSize = blockSize
        self.lib = lib
        self.popRequired = popRequired
    }

    public func hasCap(_ cap: String) -> Bool {
        caps.contains(cap)
    }
}

public struct WifiStatus: Equatable, Sendable {
    public let state: StaState
    public let failReason: FailReason?
    public let ip4: Data?
    public let ssid: Data?
    public let rssi: Int?

    public init(
        state: StaState,
        failReason: FailReason? = nil,
        ip4: Data? = nil,
        ssid: Data? = nil,
        rssi: Int? = nil
    ) {
        self.state = state
        self.failReason = failReason
        self.ip4 = ip4
        self.ssid = ssid
        self.rssi = rssi
    }
}

public struct ScanEntry: Equatable, Sendable {
    public let ssid: Data
    public let bssid: Data
    public let channel: Int
    public let rssi: Int
    public let auth: Int

    public init(ssid: Data, bssid: Data, channel: Int, rssi: Int, auth: Int) {
        self.ssid = ssid
        self.bssid = bssid
        self.channel = channel
        self.rssi = rssi
        self.auth = auth
    }
}

// MARK: - errors

/// Protocol-level error reported by the device.
public struct ProvError: Error, Equatable, CustomStringConvertible {
    public let status: Status
    public let context: String

    public init(status: Status, context: String = "") {
        self.status = status
        self.context = context
    }

    public var description: String {
        context.isEmpty ? "\(status)" : "\(context): \(status)"
    }
}

/// Response did not match the schema.
public struct DecodeError: Error, CustomStringConvertible {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var description: String { message }
}

File: android/pouchprov-core/src/main/kotlin/io/golioth/pouchprov/Messages.kt
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

package io.golioth.pouchprov

import io.golioth.pouchprov.cbor.Cbor

/**
 * Provisioning message codec (cddl/prov.cddl, protocol version 1).
 *
 * Every message is a CBOR array `[op, ...]`. This mirrors the Python `codec.py`
 * and is pinned to the device's zcbor codec via the shared golden vectors in
 * `tests/vectors/`.
 */
object Messages {
    const val PROTO_VERSION = 1

    // Reserved provisioning paths (pouch downlink/uplink entry paths).
    const val PATH_VER = ".prov/ver"
    const val PATH_AUTH = ".prov/auth"
    const val PATH_CONFIG = ".prov/config"
    const val PATH_SCAN = ".prov/scan"
    const val PATH_CRED = ".prov/cred"
    const val PATH_CTRL = ".prov/ctrl"

    // ---- .prov/ver -------------------------------------------------------

    fun encodeVerReq(): ByteArray = Cbor.encode(listOf(0L))

    fun decodeVerRsp(data: ByteArray): VersionInfo {
        val msg = decode(data, 0, "ver")
        @Suppress("UNCHECKED_CAST")
        val info = msg[2] as Map<Any?, Any?>
        return VersionInfo(
            proto = (info["proto"] as Long).toInt(),
            caps = (info["caps"] as List<*>).map { it as String },
            blockSize = (info["blk"] as Long).toInt(),
            lib = info["lib"] as String,
            popRequired = (info["pop"] as? Boolean) ?: false,
        )
    }

    // ---- .prov/auth ------------------------------------------------------

    fun encodeAuthChallenge(cliNonce: ByteArray): ByteArray {
        require(cliNonce.size == 16)
        return Cbor.encode(listOf(0L, cliNonce))
    }

    /** Returns (devNonce, devProof). */
    fun decodeAuthChallengeRsp(data: ByteArray): Pair<ByteArray, ByteArray> {
        val msg = decode(data, 0, "auth")
        val devNonce = msg[2] as ByteArray
        val devProof = msg[3] as ByteArray
        if (devNonce.size != 16 || devProof.size != 32) throw DecodeError("auth: bad nonce/proof size")
        return devNonce to devProof
    }

    fun encodeAuthProof(cliProof: ByteArray): ByteArray {
        require(cliProof.size == 32)
        return Cbor.encode(listOf(1L, cliProof))
    }

    fun decodeAuthProofRsp(data: ByteArray) {
        decode(data, 1, "auth")
    }

    // ---- .prov/config ----------------------------------------------------

    fun encodeConfigGetStatus(): ByteArray = Cbor.encode(listOf(0L))

    fun encodeConfigSet(
        ssid: ByteArray,
        password: ByteArray? = null,
        bssid: ByteArray? = null,
        channel: Int? = null,
    ): ByteArray {
        val cfg = LinkedHashMap<String, Any?>()
        cfg["ssid"] = ssid
        if (password != null) cfg["pass"] = password
        if (bssid != null) cfg["bssid"] = bssid
        if (channel != null) cfg["ch"] = channel.toLong()
        return Cbor.encode(listOf(1L, cfg))
    }

    fun encodeConfigApply(): ByteArray = Cbor.encode(listOf(2L))

    fun decodeConfigStatusRsp(data: ByteArray): WifiStatus {
        val msg = decode(data, 0, "config")
        val state = StaState.from((msg[2] as Long).toInt())
        var status = WifiStatus(state)
        if (msg.size > 3) {
            when (val detail = msg[3]) {
                is Long -> status = status.copy(failReason = FailReason.from(detail.toInt()))
                is Map<*, *> -> status = status.copy(
                    ip4 = detail["ip4"] as ByteArray,
                    ssid = detail["ssid"] as ByteArray,
                    rssi = (detail["rssi"] as? Long)?.toInt(),
                )
                else -> throw DecodeError("config: bad status detail")
            }
        }
        return status
    }

    fun decodeConfigSetRsp(data: ByteArray) {
        decode(data, 1, "config")
    }

    fun decodeConfigApplyRsp(data: ByteArray) {
        decode(data, 2, "config")
    }

    // ---- .prov/scan ------------------------------------------------------

    fun encodeScanStart(passive: Boolean? = null, periodMs: Int? = null): ByteArray {
        val params = LinkedHashMap<String, Any?>()
        if (passive != null) params["passive"] = passive
        if (periodMs != null) params["period-ms"] = periodMs.toLong()
        return Cbor.encode(listOf(0L, params))
    }

    fun encodeScanGetStatus(): ByteArray = Cbor.encode(listOf(1L))

    fun encodeScanGetResults(start: Int, count: Int): ByteArray =
        Cbor.encode(listOf(2L, start.toLong(), count.toLong()))

    fun decodeScanStartRsp(data: ByteArray) {
        decode(data, 0, "scan")
    }

    /** Returns (finished, total). */
    fun decodeScanStatusRsp(data: ByteArray): Pair<Boolean, Int> {
        val msg = decode(data, 1, "scan")
        return (msg[2] as Boolean) to (msg[3] as Long).toInt()
    }

    fun decodeScanResultsRsp(data: ByteArray): List<ScanEntry> {
        val msg = decode(data, 2, "scan")
        @Suppress("UNCHECKED_CAST")
        val list = msg[2] as List<Map<Any?, Any?>>
        return list.map { e ->
            ScanEntry(
                ssid = e["ssid"] as ByteArray,
                bssid = e["bssid"] as ByteArray,
                channel = (e["ch"] as Long).toInt(),
                rssi = (e["rssi"] as Long).toInt(),
                auth = (e["auth"] as Long).toInt(),
            )
        }
    }

    // ---- .prov/cred ------------------------------------------------------

    fun encodeCredWrite(kind: CredKind, offset: Int, total: Int, data: ByteArray): ByteArray {
        val map = LinkedHashMap<String, Any?>()
        map["kind"] = kind.code.toLong()
        map["off"] = offset.toLong()
        map["total"] = total.toLong()
        map["data"] = data
        return Cbor.encode(listOf(0L, map))
    }

    fun encodeCredFinalize(): ByteArray = Cbor.encode(listOf(1L))

    fun encodeCredGetStatus(): ByteArray = Cbor.encode(listOf(2L))

    /** Returns total bytes received for the written kind. */
    fun decodeCredWriteRsp(data: ByteArray): Int {
        val msg = decode(data, 0, "cred")
        return (msg[2] as Long).toInt()
    }

    fun decodeCredFinalizeRsp(data: ByteArray) {
        decode(data, 1, "cred")
    }

    fun decodeCredStatusRsp(data: ByteArray): Map<CredKind, Int> {
        val msg = decode(data, 2, "cred")
        @Suppress("UNCHECKED_CAST")
        val map = msg[2] as Map<Any?, Any?>
        return map.entries.associate { (k, v) -> CredKind.from((k as String).toInt()) to (v as Long).toInt() }
    }

    // ---- .prov/ctrl ------------------------------------------------------

    fun encodeCtrl(op: CtrlOp): ByteArray = Cbor.encode(listOf(op.code.toLong()))

    fun decodeCtrlRsp(data: ByteArray, op: CtrlOp) {
        decode(data, op.code, "ctrl")
    }

    // ---- shared ----------------------------------------------------------

    private fun decode(data: ByteArray, expectOp: Int, context: String): List<Any?> {
        val msg = try {
            Cbor.decode(data)
        } catch (e: Exception) {
            throw DecodeError("$context: bad CBOR: ${e.message}")
        }
        if (msg !is List<*> || msg.size < 2) throw DecodeError("$context: not a response array")
        if ((msg[0] as? Long)?.toInt() != expectOp) throw DecodeError("$context: op ${msg[0]}, expected $expectOp")
        val status = Status.from((msg[1] as Long).toInt())
        if (status != Status.OK) throw ProvError(status, context)
        return msg
    }
}

// ---- enums ---------------------------------------------------------------

enum class Status(val code: Int) {
    OK(0),
    INVALID_PROTO(1),
    INVALID_ARGUMENT(2),
    INTERNAL_ERROR(3),
    UNAUTHORIZED(4),
    INVALID_STATE(5),
    BUSY(6),
    ;

    companion object {
        fun from(code: Int): Status = entries.firstOrNull { it.code == code }
            ?: throw DecodeError("unknown status $code")
    }
}

enum class StaState(val code: Int) {
    CONNECTED(0),
    CONNECTING(1),
    DISCONNECTED(2),
    FAILED(3),
    ;

    companion object {
        fun from(code: Int): StaState = entries.firstOrNull { it.code == code }
            ?: throw DecodeError("unknown sta-state $code")
    }
}

enum class FailReason(val code: Int) {
    AUTH_ERROR(0),
    NETWORK_NOT_FOUND(1),
    ;

    companion object {
        fun from(code: Int): FailReason = entries.firstOrNull { it.code == code }
            ?: throw DecodeError("unknown fail-reason $code")
    }
}

enum class CredKind(val code: Int) {
    DEVICE_CERT(0),
    PRIVATE_KEY(1),
    CA_CERT(2),
    ;

    companion object {
        fun from(code: Int): CredKind = entries.firstOrNull { it.code == code }
            ?: throw DecodeError("unknown cred kind $code")
    }
}

enum class CtrlOp(val code: Int) {
    RESET(0),
    REPROVISION(1),
    END(2),
}

// ---- value types ---------------------------------------------------------

data class VersionInfo(
    val proto: Int,
    val caps: List<String>,
    val blockSize: Int,
    val lib: String,
    val popRequired: Boolean,
) {
    fun hasCap(cap: String): Boolean = cap in caps
}

data class WifiStatus(
    val state: StaState,
    val failReason: FailReason? = null,
    val ip4: ByteArray? = null,
    val ssid: ByteArray? = null,
    val rssi: Int? = null,
) {
    override fun equals(other: Any?): Boolean =
        other is WifiStatus && state == other.state && failReason == other.failReason &&
            (ip4?.contentEquals(other.ip4) ?: (other.ip4 == null)) &&
            (ssid?.contentEquals(other.ssid) ?: (other.ssid == null)) && rssi == other.rssi

    override fun hashCode(): Int = state.hashCode()
}

data class ScanEntry(
    val ssid: ByteArray,
    val bssid: ByteArray,
    val channel: Int,
    val rssi: Int,
    val auth: Int,
) {
    val authName: String get() = securityName(auth)

    override fun equals(other: Any?): Boolean =
        other is ScanEntry && ssid.contentEquals(other.ssid) && bssid.contentEquals(other.bssid) &&
            channel == other.channel && rssi == other.rssi && auth == other.auth

    override fun hashCode(): Int = ssid.contentHashCode()
}

/** Human-readable name for a Wi-Fi security type (Zephyr enum wifi_security_type). */
fun securityName(auth: Int): String = when (auth) {
    0 -> "Open"
    1 -> "WPA2-PSK"
    2 -> "WPA2-PSK-SHA256"
    3 -> "WPA3-SAE"
    4 -> "WPA3-SAE-H2E"
    5 -> "WPA3-SAE-AUTO"
    6 -> "WAPI"
    7 -> "EAP-TLS"
    8 -> "WEP"
    9 -> "WPA-PSK"
    10 -> "WPA/WPA2-Auto"
    11 -> "DPP"
    else -> "unknown($auth)"
}

// ---- exceptions ----------------------------------------------------------

/** Protocol-level error reported by the device. */
class ProvError(val status: Status, context: String = "") :
    Exception(if (context.isNotEmpty()) "$context: $status" else status.toString())

/** Response did not match the schema. */
class DecodeError(message: String) : Exception(message)
